### DISCUSSION TITLE
Issues with Jacobian incorporating second time derivative

### DISCUSSION THREAD

**Discussion by NateAM** [idaholab/moose#25768] [Q&A General] [answered]
I have a kernel where the residual is defined as

    Real linearMomentumInertia::computeQpResidual( ){

        Real result = 0;

        assembleRequiredTerms( );

        result -= _test[ _i ][ _qp ] * ( _density_dot[ _qp ] * _velocity[ _index ] + _density[ _qp ] * _acceleration[ _index ] );

        return result;

    }

where

    void linearMomentumInertia::assembleRequiredTerms( ){
        /*!
         * Assemble the required terms for computation of the residual and Jacobians
         */

        if ( _is_displacement ){

            _velocity[ 0 ] = _v1_dot[ _qp ];
            _velocity[ 1 ] = _v2_dot[ _qp ];
            _velocity[ 2 ] = _v3_dot[ _qp ];

            _velocity_factors[ 0 ] = _dv1_dot_dv1[ _qp ];
            _velocity_factors[ 1 ] = _dv2_dot_dv2[ _qp ];
            _velocity_factors[ 2 ] = _dv3_dot_dv3[ _qp ];

            _acceleration[ 0 ] = _v1_dotdot[ _qp ];
            _acceleration[ 1 ] = _v2_dotdot[ _qp ];
            _acceleration[ 2 ] = _v3_dotdot[ _qp ];

            _acceleration_factors[ 0 ] = _dv1_dotdot_dv1[ _qp ];
            _acceleration_factors[ 1 ] = _dv2_dotdot_dv2[ _qp ];
            _acceleration_factors[ 2 ] = _dv3_dotdot_dv3[ _qp ];

        }
        else{

            _velocity[ 0 ] = _v1[ _qp ];
            _velocity[ 1 ] = _v2[ _qp ];
            _velocity[ 2 ] = _v3[ _qp ];

            _velocity_factors[ 0 ] = 1;
            _velocity_factors[ 1 ] = 1;
            _velocity_factors[ 2 ] = 1;

            _acceleration[ 0 ] = _v1_dot[ _qp ];
            _acceleration[ 1 ] = _v2_dot[ _qp ];
            _acceleration[ 2 ] = _v3_dot[ _qp ];

            _acceleration_factors[ 0 ] = _dv1_dot_dv1[ _qp ];
            _acceleration_factors[ 1 ] = _dv2_dot_dv2[ _qp ];
            _acceleration_factors[ 2 ] = _dv3_dot_dv3[ _qp ];

        }

    }

and I define the jacobians via

    Real linearMomentumInertia::computeQpJacobian( ){

        // Add the velocity terms
        Real result = 0;

        assembleRequiredTerms( );

        result -= _test[ _i ][ _qp ] * ( _density_dot[ _qp ] * _velocity_factors[ _index ] + _density[ _qp ] * _acceleration_factors[ _index ] ) * _phi[ _j ][ _qp ];

        return result + computeQpOffDiagJacobian( _var.number( ) );

    }

    Real linearMomentumInertia::computeQpOffDiagJacobian( unsigned int jvar ){

        Real result = 0;

        assembleRequiredTerms( );

        // Add the Jacobians w.r.t. the density
        if ( _density_ind == jvar ){

            result -= _test[ _i ][ _qp ] * ( _ddensity_dot_ddensity[ _qp ] * _velocity[ _index ] + _acceleration[ _index ] ) * _phi[ _j ][ _qp ];

        }

        if ( _use_displaced_mesh ){

            // Add the Jacobians w.r.t. the mesh displacement
            if ( ( _displacements.size( ) > 0 ) && ( jvar == _displacements[ 0 ] ) ){

                result += computeQpResidual( ) * _grad_phi[ _j ][ _qp ]( 0 );

            }

            else if ( ( _displacements.size( ) > 1 ) && ( jvar == _displacements[ 1 ] ) ){

                result += computeQpResidual( ) * _grad_phi[ _j ][ _qp ]( 1 );

            }

            else if ( ( _displacements.size( ) > 2 ) && ( jvar == _displacements[ 2 ] ) ){

                result += computeQpResidual( ) * _grad_phi[ _j ][ _qp ]( 2 );

            }

        }

        return result;

    }

I am getting access to `coupledDotDot` by requring the use of the Newmark Beta timestepper rather than using `setUDotDotRequested` locally because I was having trouble getting the answer in https://github.com/idaholab/moose/discussions/25300 to work.

If I hard code `_acceleration` so that it is a constant, and set the `_dv1_dot_dv1`, `_dv2_dot_dv2`, and `_dv3_dot_dv3` terms to zero and run with `use_displaced_mesh = true`, a 1D run with the PETSc options `-snes_test_jacobian -snes_test_jacobian_view` shows no issue with the Jacobian. If I run with the accelerations turned on but with `use_displaced_mesh = false` then I get a similarly good Jacobian. However, if I use the solved for acceleration and use the displaced mesh, I get noticeable errors (relative error of order 7e-7) in the Jacobian.

Is there something obviously wrong with my Jacobian? If PETSc is only using forward difference can I force it to use central differences for a higher quality check of the Jacobian?

**Comment by lindsayad** [comment, marked as answer]
I would not call 7e-7 significant error. Have you tried comparing nonlinear iteration counts with `solve_type = NEWTON` vs. `solve_type = PJFNK`? I would only really be concerned about the Jacobian quality if you're seeing more nonlinear iterations with NEWTON than with PJFNK

**Reply by NateAM** [reply]
> I would not call 7e-7 significant error.

Okay, that's good to know! I had seen some comments about relative magnitudes greater than 1e-7 being an issue

> Have you tried comparing nonlinear iteration counts with solve_type = NEWTON vs. solve_type = PJFNK? I would only really be concerned about the Jacobian quality if you're seeing more nonlinear iterations with NEWTON than with PJFNK

I have looked at that `NEWTON` with a `LU` preconditioner seems to be really good so I think I was probably making a mountain out of a molehill.

**Reply by lindsayad** [reply]
I'm usually pretty happy up into the 1e-6 to 1e-5 range. If you have a "noisy" residual, then sometimes you can get better matching with PETSc if you pass the `-snes_test_err` option which controls the finite difference size. The default is ~1e-8. Sometimes I get better with tighter (e.g. `-snes_test_err 1e-9`), sometimes with looser (e.g. `-snes_test_err 1e-6`)
